lambdas/package/index.py

```diff
@@ -0,0 +1,186 @@
+}
+
+
+class IncompleteResultException(Exception):
+    """
+    Exception indicating an incomplete response
+    (e.g., from S3 Select)
+    """
+
+
+def buffer_s3response(s3response):
+    """
+    Read a streaming response from s3 select into a
+    StringIO buffer
+    """
+    response = io.StringIO()
+    end_event_received = False
+    stats = None
+    for event in s3response['Payload']:
+        if 'Records' in event:
+            records = event['Records']['Payload'].decode()
+            response.write(records)
+        elif 'Progress' in event:
+            print(event['Progress']['Details'])
+        elif 'Stats' in event:
+            print(stats)
+        elif 'End' in event:
+            # End event indicates that the request finished successfully
+            end_event_received = True
+
+    if not end_event_received:
+        raise IncompleteResultException("Error: Received an incomplete response from S3 Select.")
+    response.seek(0)
+    return response
+
+
+def get_logical_key_folder_view(df):
+    """
+    Post process a set of logical keys to return only the
+    top-level folder view (a special case of the s3-select
+    lambda).
+    """
+    # matches all strings; everything before and including the first
+    # / is extracted
+    folder = pd.Series(df.logical_key.dropna().str.extract('([^/]+/?).*')[0].unique())
+    return dict(
+        prefixes=folder[folder.str.endswith('/')].sort_values().tolist(),
+        objects=folder[~folder.str.endswith('/')].sort_values().tolist()
+    )
+
+
+def get_s3_client(aws_access_key_id, aws_secret_access_key, aws_session_token):
+    """
+    Create an S3 Client using the provided credentials
+    """
+    session = boto3.Session(
+        aws_access_key_id=aws_access_key_id,
+        aws_secret_access_key=aws_secret_access_key,
+        aws_session_token=aws_session_token
+    )
+    s3_client = session.client('s3')
+    return s3_client
+
+
+def call_s3_select(s3_client, bucket, key, logical_key_prefix, detail=False):
+    """
+    Call S3 Select to read only the logical keys from a
+    package manifest that match the desired folder path
+    prefix
+    """
+
+    if logical_key_prefix is None:
+        logical_key_prefix = ""
+
+    prefix_length = len(logical_key_prefix)
+    sanitized = logical_key_prefix.replace("'", "''")
+
+    if detail:
+        logical_key = sanitized
+        sql_stmt = f"SELECT s.* FROM s3object s WHERE s.logical_key = '{logical_key}' LIMIT 1"
+    else:
+        prefix = sanitized
+        sql_stmt = f"SELECT SUBSTRING(s.logical_key, {prefix_length + 1}) AS logical_key FROM s3object s"
+        if prefix:
+            sql_stmt += f" WHERE SUBSTRING(s.logical_key, 1, {prefix_length}) = '{prefix}'"
+
+    print(sql_stmt)
+    response = s3_client.select_object_content(
+        Bucket=bucket,
+        Key=key,
+        ExpressionType='SQL',
+        Expression=sql_stmt,
+        InputSerialization={
+            'JSON': {'Type': 'DOCUMENT'},
```

Comment:
This should be `'Type': LINES`? I'm really confused how this even works because you're querying a manifest.

Reply:
Seems to work fine either way, but I'll change it.